### PR TITLE
Implement file system extensions and ext point definitions

### PR DESCRIPTION
**Pull request by loganga**

### CODE REVIEW

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		this.ss = CoreActivator.getSessionStorage();
+		this.fc = new FileController(ss);
+		this.extMgr = FileSystemExtensionManager.getInstance();
+	}
+	
+	@Override
+	public void fileCreated(long fileID, String name, Path absolutePath, String projectRelativePath, long projectID) {		
+		// NOTE: This is only here for now so that I don't forget to include it when making extensions in the model plugin
+		File f = new File(fileID, name, projectRelativePath, 1, ss.getUsername(), new Date().toString()); // doing date like this is questionable, but it's not being used
+		ss.setFile(projectID, f);
```
Comment (fahslaj) [inline org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java:37]: Shouldn't this be done in the models plugin?

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		Set<ICoreExtension> extensions = extMgr.getExtensions(APIExtensionIDs.FILE_CREATE_ID);
+		for (ICoreExtension e : extensions) {
+			IFSFileCreateExt createExt = (IFSFileCreateExt) e;
+			createExt.fileCreated(f, iFile);
+		}
+	}
+
+	@Override
+	public void fileCreateFailed(String fileName, Path absolutePath, long projectID, byte[] fileBytes) {
+
```
Comment (fahslaj) [inline org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java]: Should we do something here? If not, just add a comment saying something like `// Do nothing`

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileMove.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+public class FSFileMove implements IFileMoveResponse {
+
+	@Override
+	public void fileMoved(long fileID, Path newWorkspaceRelativePath) {
+		new Thread(APIFactory.createFilePullDiffSendChanges(fileID)).start();
+	}
+
+	@Override
+	public void fileMoveFailed(long fileID, Path newWorkspaceRelativePath) {
+
```
Comment (fahslaj) [inline org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileMove.java]: I believe you need to try to undo the file move here. The only time this function will be called is if the user moves a file on the filesystem, the directorywatching detects it, and we send a filemoverequest. Then this response extension will get run when the move fails, so we need to move it back. This might require sending in the old path as the parameter.

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileMove.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import java.nio.file.Path;
+
+import org.code.toboggan.core.api.APIFactory;
+import org.code.toboggan.network.request.extensionpoints.file.IFileMoveResponse;
+
+public class FSFileMove implements IFileMoveResponse {
+
+	@Override
+	public void fileMoved(long fileID, Path newWorkspaceRelativePath) {
```
Comment (fahslaj) [inline org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileMove.java]: will this parameter #2 actually be a workspace relative path? or will it be absolute? If it will be absolute, please rename the variable. Same with the function below.

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileRename.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.nio.file.Path;
+
+import org.code.toboggan.core.api.APIFactory;
+import org.code.toboggan.network.request.extensionpoints.file.IFileRenameResponse;
+
+public class FSFileRename implements IFileRenameResponse {
+
+	@Override
+	public void fileRenamed(long fileID, Path newWorkspaceRelativePath, String newName) {
```
Comment (fahslaj) [inline org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileRename.java]: Same comment about workspace relative path vs absolute path

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/project/FSProjectCreate.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+	@Override
+	public void projectCreationFailed(String name) {
+		// TODO Auto-generated method stub
+		
+	}
+
+	@Override
+	public void subscribed(long projectId) {
+		// TODO Auto-generated method stub
```
Comment (fahslaj) [inline org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/project/FSProjectCreate.java]: remove the `TODO Auto-generated` comment and replace it with a `// Do nothing` if we truely don't need to implement these functions

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/util/FSUtils.java [hunk 7]
```diff
@@ -36,4 +36,9 @@ public static Path getLocationForRelativePath(Path workspaceRelativePath) {
 		return result;
 	}
 	
+//	public static Path getLocationForProjectRelativePath(Path projectRelativePath) {
```
Comment (fahslaj) [inline org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/util/FSUtils.java]: Why is this here and commented out

**Review by fahslaj — CHANGES_REQUESTED** [review]

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java (hunk 1 above)
Comment (loganga) [reply org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java:37]: There's a comment above it that explains that why it's here and not in the models plugin, but to further elaborate, the way this is done in the original plugin is that the project creation call waits with a Semaphore for all of the files to finish and then does a GetFiles request to create the file metadata and set it for the project metadata all at once. The reason I put this in here was because I wanted to make sure that I would remember to actually create the file models when transferring model operations for everything else to the plugin because in the old plugin, this was intertwined with the project creation code in a weird way.

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/util/FSUtils.java (hunk 7 above)
Comment (loganga) [reply org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/util/FSUtils.java]: Oh, my bad, I forgot to remove that

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileMove.java (hunk 4 above)
Comment (loganga) [reply org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileMove.java]: Should be the absolute path, I'll fix the name.

File: org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java (hunk 1 above)
Comment (fahslaj) [reply org.code.toboggan.filesystem/src/org/code/toboggan/filesystem/extensions/file/FSFileCreate.java:37]: Ok. That makes sense.

**Review by fahslaj — APPROVED** [review]

**Comment by fahslaj** [comment]
Looks good.
